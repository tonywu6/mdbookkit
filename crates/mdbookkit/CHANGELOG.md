# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0](https://github.com/tonywu6/mdbookkit/releases/tag/mdbookkit-v0.1.0) - 2025-04-03

### Other

- [link-forever] fix(ci): page order in snapshots
- [link-forever] fix: images in links
- [docs] touch up crate docs
- [docs] finalize book
- [docs] socials
- [link-forever] rewrite absolute paths to relative paths
- [link-forever] check canonical URLs
- [rustdoc-link] try to fix spurious RA progress reports
- [ci] debug flaky test
- [rustdoc-link] make RA probing less flaky?
- [ci] setup actions
- [docs] bundle fonts
- [link-forever] improve fallback for new repos
- [link-forever] docs
- [link-forever] support images
- link forever
- link check
- clean up
- make into one crate + features
