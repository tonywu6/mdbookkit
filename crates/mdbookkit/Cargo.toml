[package]
name = "mdbookkit"
publish = true
version = "1.0.0"

edition.workspace = true
rust-version.workspace = true

authors.workspace = true
license.workspace = true
repository.workspace = true

categories = ["command-line-utilities"]
description = "toolkit for mdBook ðŸ“–"
documentation = "https://tonywu6.github.io/mdbookkit/"
keywords = ["mdbook", "documentation", "rust-analyzer", "lsp", "permalink"]
readme = "README.md"

autobins = false
autotests = false

[dependencies]
anyhow = { workspace = true }
async-lsp = { version = "0.2.2", optional = true }
cargo_toml = { version = "0.21.0", optional = true }
chrono = { version = "0.4.40", features = [
  "clock",
], default-features = false, optional = true }
clap = { workspace = true, optional = true }
console = { version = "0.15.11", optional = true }
dirs = { version = "6.0.0", optional = true }
env_logger = { workspace = true, optional = true }
git2 = { version = "0.20.1", optional = true }
gix-url = { version = "0.29.0", optional = true }
indicatif = { version = "0.17.11", optional = true }
log = { workspace = true }
lsp-types = { version = "0.95.0", optional = true }
mdbook = { workspace = true, optional = true }
miette = { workspace = true, optional = true }
owo-colors = { version = "4.2.0", optional = true }
percent-encoding = { version = "2.3.1", optional = true }
proc-macro2 = { version = "1.0.94", features = [
  "span-locations",
], optional = true }
pulldown-cmark = { workspace = true }
pulldown-cmark-to-cmark = { workspace = true }
serde = { workspace = true, features = ["rc"] }
serde_json = { workspace = true }
sha2 = { version = "0.10.8", optional = true }
shlex = { workspace = true, optional = true }
syn = { version = "2.0.99", optional = true }
tap = { workspace = true }
tempfile = { version = "3.18.0", optional = true }
tokio = { workspace = true, optional = true }
tokio-util = { version = "0.7.13", features = ["compat"], optional = true }
toml = { workspace = true, optional = true }
tower = { version = "0.5.2", optional = true }
url = { workspace = true, features = ["serde"], optional = true }

[dev-dependencies]
insta = { workspace = true }
similar = { workspace = true }
util-testing = { workspace = true }

[features]
lib-rustdoc-link = [
  "dep:async-lsp",
  "dep:cargo_toml",
  "dep:dirs",
  "dep:lsp-types",
  "dep:proc-macro2",
  "dep:sha2",
  "dep:shlex",
  "dep:syn",
  "dep:tempfile",
  "dep:tokio",
  "dep:tokio",
  "dep:tokio-util",
  "dep:tower",
  "tokio/process",
  "tokio/rt",
  "tokio/time",
]
rustdoc-link = [
  "lib-rustdoc-link",
  "dep:toml",
  "tokio/fs",
  "tokio/rt-multi-thread",
  "common-cli",
  "common-logger",
]

lib-link-forever = ["dep:mdbook", "dep:percent-encoding", "dep:url"]
link-forever = [
  "lib-link-forever",
  "dep:git2",
  "dep:gix-url",
  "common-cli",
  "common-logger",
]

common-logger = [
  "dep:chrono",
  "dep:console",
  "dep:env_logger",
  "dep:indicatif",
  "dep:miette",
  "dep:owo-colors",
]

common-cli = ["dep:clap", "dep:mdbook", "dep:toml"]

default = []

# aliases
mdbook-link-forever = ["link-forever"]
mdbook-rustdoc-link = ["rustdoc-link"]

[[bin]]
name = "mdbook-rustdoc-link"
path = "src/bin/rustdoc_link/main.rs"
required-features = ["rustdoc-link"]

[[test]]
name = "rustdoc-link"
path = "tests/rustdoc_link.rs"
required-features = ["rustdoc-link"]

[[bin]]
name = "mdbook-link-forever"
path = "src/bin/link_forever/main.rs"
required-features = ["link-forever"]

[[test]]
name = "link-forever"
path = "tests/link_forever.rs"
required-features = ["link-forever"]

[package.metadata.docs.rs]
all-features = true
