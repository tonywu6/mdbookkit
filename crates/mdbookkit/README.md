# mdbookkit

![mdbookkit hero image](/docs/src/media/banner.webp)

Quality-of-life plugins for your [mdBook] project.

Right now, there are two mdBook [preprocessors], both for generating **correct and
versioned** external links from **easy-to-write** markup.

- [**`mdbook-rustdoc-link`**](https://tonywu6.github.io/mdbookkit/rustdoc-link)

  _rustdoc_-style linking for Rust APIs: write types and function names, get links to
  _docs.rs_

- [**`mdbook-link-forever`**](https://tonywu6.github.io/mdbookkit/link-forever)

  _Permalinks_ for your source tree: write relative paths, get links to GitHub.

> [!TIP]
>
> Preprocessors are standalone programs that mdBook invokes to transform your Markdown
> sources before rendering them.

## Installation

If you are interested in any of these plugins, visit their respective pages for usage
instructions, linked above.

If you want to install all of them:

```bash
cargo install mdbookkit --all-features
```

Precompiled binaries are also available from [GitHub releases][gh-releases].

## License

This project is released under the [Apache 2.0 License](/LICENSE-APACHE.md) and the
[MIT License](/LICENSE-MIT.md).

<!-- prettier-ignore-start -->

[mdBook]: https://rust-lang.github.io/mdBook/
[`mdbookkit`]: https://crates.io/crates/mdbookkit
[preprocessors]: https://rust-lang.github.io/mdBook/format/configuration/preprocessors.html
[gh-releases]: https://github.com/tonywu6/mdbookkit/releases

<!-- prettier-ignore-end -->
