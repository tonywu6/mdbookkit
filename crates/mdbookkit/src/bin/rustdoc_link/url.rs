use std::path::PathBuf;

use anyhow::{bail, Result};
use lsp_types::Url;

/// [`Url::to_file_path()`] with an actual [`std::error::Error`].
pub trait UrlToPath {
    fn to_path(&self) -> Result<PathBuf>;
}

impl UrlToPath for Url {
    fn to_path(&self) -> Result<PathBuf> {
        match self.to_file_path() {
            Ok(path) => Ok(path),
            Err(()) => bail!("failed to convert {self} to a file path"),
        }
    }
}
