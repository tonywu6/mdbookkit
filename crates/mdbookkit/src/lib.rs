#[cfg(feature = "common-logger")]
pub mod diagnostics;
pub mod logging;
pub mod markdown;

pub mod bin;
