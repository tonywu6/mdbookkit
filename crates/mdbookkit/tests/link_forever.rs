use anyhow::Result;
use log::LevelFilter;
use tap::Pipe;

use url::Url;

use mdbookkit::{
    bin::link_forever::{Config, Environment, GitHubPermalink, Pages},
    logging::ConsoleLogger,
    markdown::mdbook_markdown,
};
use util_testing::{portable_snapshots, test_document, CARGO_WORKSPACE_DIR};

#[test]
fn test_links() -> Result<()> {
    ConsoleLogger::install("link-forever");

    let env = Environment {
        book_src: CARGO_WORKSPACE_DIR.join("crates/mdbookkit/")?,
        vcs_root: CARGO_WORKSPACE_DIR.clone(),
        fmt_link: GitHubPermalink::new("lorem", "ipsum", "dolor").pipe(Box::new),
        markdown: mdbook_markdown(),
        config: Config {
            book_url: Some("https://example.org/book".parse::<Url>()?.into()),
            ..Default::default()
        },
    };

    let mut pages = Pages::new(mdbook_markdown());

    let tests = [
        test_document!("tests/ra-known-quirks.md"), // only for providing anchors
        test_document!("tests/link-forever.md"),
        test_document!("../README.md"),
    ];

    for page in tests.iter() {
        pages.insert(page.file.clone(), page.source)?;
    }

    env.resolve(&mut pages);

    for page in tests {
        let output = pages.emit(&page.file)?;
        let name = page.name;
        portable_snapshots!().test(|| insta::assert_snapshot!(format!("{name}"), output))?;
    }

    let report = env
        .report(&pages)
        .level(LevelFilter::Debug)
        .names(|url| env.rel_path(url))
        .colored(false)
        .logging(false)
        .build()
        .to_report();

    portable_snapshots!().test(|| insta::assert_snapshot!("_stderr", report))?;

    Ok(())
}
