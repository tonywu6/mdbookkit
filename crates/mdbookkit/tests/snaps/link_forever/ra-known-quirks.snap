---
source: crates/mdbookkit/tests/link_forever.rs
expression: output
---
# primitive types resolved to nightly

- [u8]
- [u32]
- [f64]
- [char]
- [str]

# associated items on primitive types

- [str::parse]
- [f64::MIN_POSITIVE]

# macro_export

- [std::vec!]
- [std::format!]
- [tokio::main!]
